### PR TITLE
mapnik: unstable-2023-11-28 -> 4.0.0

### PR DESCRIPTION
**Pull request by hummeltech**
## Description of changes

* Removed no-longer-needed patches:
  * `datasource-ogr-test-should-fail.patch`: Fixed [here](https://github.com/mapnik/mapnik/commit/2b932be0ca0102dd4c790e3b48a2f76532a22b89)
* Use `sparsehash` package rather than bundled `sparsehash`
  * By introducing a patch (`use-sparsehash-package.patch`)
  * Requires adding `sparsehash` as dependency to `python-mapnik`
* Use `protozero` package rather than bundled `protozero`
  * By enabling `USE_EXTERNAL_MAPBOX_PROTOZERO`
* Use `std::filesystem` rather than `boost:filesystem`: Available since [here](https://github.com/mapnik/mapnik/commit/338942b490d22d6a58745259dfffdb3084f11aa7)
  * This is now the default
  * Requires adding `boost` as dependency to `Tirex` as it is no longer inherited from `Mapnik`
    * Along with an [upstream patch](https://github.com/openstreetmap/tirex/commit/137903be9b7b35dde4c7010e65faa16bcf6ad476) to properly link to it
* Fixed `macOS`/`darwin` builds
  * By disabling `USE_MEMORY_MAPPED_FILE` (only when `stdenv.isDarwin`)
* In addition to not building the `demo viewer` & in order to save resources, stop building the following dev-related targets:
  * The `demo c++ application`
    * By disabling `BUILD_DEMO_CPP`
  * The `benchmark project`
    * By disabling `BUILD_BENCHMARK`
    
Also:
* maintainers: corrected email address
* apacheHttpdPackages.mod_tile: Support for Mapnik >= v4.0.0-rc2
  * Requires a patch to use `std::optional` rather than `boost::optional`: Change made [here](https://github.com/mapnik/mapnik/commit/d7a490ffc2aed50b02e33e388b1a9d57d8a1962c)
* perlPackages.Tirex: Support for Mapnik v4.0.0
* python311Packages.python-mapnik: unstable-2023-02-23 -> unstable-2024-02-22
  * Requires a patch to use `std::optional` rather than `boost::optional`: Change made [here](https://github.com/mapnik/mapnik/commit/d7a490ffc2aed50b02e33e388b1a9d57d8a1962c)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by marcusramberg** [comment]
Result of `nixpkgs-review pr 304866` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile</li>
    <li>mapnik</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
  </ul>
</details>

**Review by marcusramberg — APPROVED** [review]
lgtm.

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 304866` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>mapnik</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
  </ul>
</details>

**Comment by hummeltech** [comment]
@sikmir, thank you for the heads up, I have added 61ea447e4763428f106bebde48305f5c14a82b55 in order to address the failures.

**Comment by sikmir** [comment]
> @sikmir, thank you for the heads up, I have added [61ea447](https://github.com/NixOS/nixpkgs/commit/61ea447e4763428f106bebde48305f5c14a82b55) in order to address the failures.

```sh
       > =========================== short test summary info ============================
       > FAILED test/python_tests/cairo_test.py::test_passing_pycairo_context_png - FileNotFoundError: [Errno 2] No such file or directory: 'images/pycairo/cai...
       > FAILED test/python_tests/cairo_test.py::test_pycairo_svg_surface1 - RuntimeError: Could not load map file: File does not exist of '../data/good...
       > FAILED test/python_tests/cairo_test.py::test_pycairo_svg_surface2 - RuntimeError: Could not load map file: File does not exist of '../data/good...
       > FAILED test/python_tests/cairo_test.py::test_pycairo_svg_surface3 - RuntimeError: Could not load map file: File does not exist of '../data/good...
       > FAILED test/python_tests/cairo_test.py::test_pycairo_pdf_surface1 - RuntimeError: Could not load map file: File does not exist of '../data/good...
       > FAILED test/python_tests/cairo_test.py::test_pycairo_pdf_surface2 - RuntimeError: Could not load map file: File does not exist of '../data/good...
       > FAILED test/python_tests/cairo_test.py::test_pycairo_pdf_surface3 - RuntimeError: Could not load map file: File does not exist of '../data/good...
       > ====== 7 failed, 346 passed, 7 deselected, 1 warning in 65.77s (0:01:05) =======
```

**Comment by marcusramberg** [comment]
> > @sikmir, thank you for the heads up, I have added [61ea447](https://github.com/NixOS/nixpkgs/commit/61ea447e4763428f106bebde48305f5c14a82b55) in order to address the failures.
> 
> ```shell
>        > =========================== short test summary info ============================
>        > FAILED test/python_tests/cairo_test.py::test_passing_pycairo_context_png - FileNotFoundError: [Errno 2] No such file or directory: 'images/pycairo/cai...
>        > FAILED test/python_tests/cairo_test.py::test_pycairo_svg_surface1 - RuntimeError: Could not load map file: File does not exist of '../data/good...
>        > FAILED test/python_tests/cairo_test.py::test_pycairo_svg_surface2 - RuntimeError: Could not load map file: File does not exist of '../data/good...
>        > FAILED test/python_tests/cairo_test.py::test_pycairo_svg_surface3 - RuntimeError: Could not load map file: File does not exist of '../data/good...
>        > FAILED test/python_tests/cairo_test.py::test_pycairo_pdf_surface1 - RuntimeError: Could not load map file: File does not exist of '../data/good...
>        > FAILED test/python_tests/cairo_test.py::test_pycairo_pdf_surface2 - RuntimeError: Could not load map file: File does not exist of '../data/good...
>        > FAILED test/python_tests/cairo_test.py::test_pycairo_pdf_surface3 - RuntimeError: Could not load map file: File does not exist of '../data/good...
>        > ====== 7 failed, 346 passed, 7 deselected, 1 warning in 65.77s (0:01:05) =======
> ```

Also seeing this on aarch64-darwin. Also the commit message of the fix (once it's actually fixed :) needs to be updated.

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 304866` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>mapnik</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
  </ul>
</details>

File: pkgs/development/libraries/mapnik/default.nix [hunk 1]
```diff
@@ -83,21 +86,32 @@ stdenv.mkDerivation rec {
...
+    # Save time by not building some development-related code.
+    (lib.cmakeBool "BUILD_BENCHMARK" false)
+    (lib.cmakeBool "BUILD_DEMO_CPP" false)
+    ## Would require QT otherwise.
+    (lib.cmakeBool "BUILD_DEMO_VIEWER" false)
+    # Use 'protozero' package.
+    (lib.cmakeBool "USE_EXTERNAL_MAPBOX_PROTOZERO" true)
+  ] ++ lib.optionals stdenv.isDarwin [
+    # macOS builds fail when using memory mapped file cache.
+    (lib.cmakeBool "USE_MEMORY_MAPPED_FILE" false)
```
Comment (sikmir) [inline pkgs/development/libraries/mapnik/default.nix:103]:
```suggestion
    # macOS builds fail when using memory mapped file cache.
    (lib.cmakeBool "USE_MEMORY_MAPPED_FILE" (!stdenv.isDarwin))
```

File: pkgs/development/python-modules/python-mapnik/default.nix [hunk 2]
```diff
@@ -23,21 +23,20 @@
...
   pytestCheckHook,
-  stdenv,
+  darwin,
+  sparsehash,
 }:
 
 buildPythonPackage rec {
   pname = "python-mapnik";
-  version = "unstable-2023-02-23";
+  version = "unstable-2024-02-22";
```
Comment (sikmir) [inline pkgs/development/python-modules/python-mapnik/default.nix:32]:
```suggestion
  version = "3.0.16-unstable-2024-02-22";
```

File: pkgs/development/libraries/mapnik/default.nix (hunk 1 above)
Comment (hummeltech) [reply pkgs/development/libraries/mapnik/default.nix]: Thanks @sikmir, good suggestion!

**Pull request merged**